### PR TITLE
[SPARK-35598] Improve Spark-ML PCA analysis

### PR DESCRIPTION
**Pull request by Antozamm**
### What changes were proposed in this pull request?

When performing a PCA, the covariance matrix is calculate first.
In the function RowMAtrix.computeCovariance() there is the following code:

```
    if (rows.first().isInstanceOf[DenseVector]) {
      computeDenseVectorCovariance(mean, n, m)
    } else {
      computeSparseVectorCovariance(mean, n, m)
    }
```
As can be seen, there are two different function to compute the covariance matrix depending on whether the matrix is sparse or dense. Now, the decision if the matrix is sparse or dense is based only on the type of the first row (DenseVector or SparseVector).

I propose to calculate the **sparsity** of the matrix, as the ratio between number-of-zeros and total-number-of elements. And use the sparsity to switch between the two functions.

In later pull request #32734 the code checks that all rows are SparseVector, if all rows are sparse it uses the sparse covariance computation. 
Still, I would prefer to have the switch based on the actual sparsity of the matrix, rather than on its implementation. To notice you can have 2 matrices, with same elements and same number of zeros, implemented with SparseVector or DenseVector.

### Why are the changes needed?

Have more consistent results. With my implementation 2 matrices with same elements produce same results, independently if they are implemented using SparseVector or using DenseVector

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

I added 2 tests in the RowMatrixSuite.scala to verify that the sparsity is calculated properly.

### CODE REVIEW

**Comment by Antozamm** [comment]
Sure it can be verified performing a PCA analysis, the one in the doc, for example.
There are also PCA test in the RowMatrixSuite

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala [hunk 1]
```diff
@@ -439,13 +438,27 @@ class RowMatrix @Since("1.0.0") (
...
       computeSparseVectorCovariance(mean, n, m)
     }
   }
 
+  /**
+   * Calculate percent sparsity of the matrix as number-of-zero-elements
+   * divided by the total number of elements of the matrix
+   */
+  @Since("3.2.0")
+  def calcSparsity(): Double = rows.map{
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala:456]: This should be private

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala [hunk 2]
```diff
@@ -439,13 +438,27 @@ class RowMatrix @Since("1.0.0") (
       "  Cannot compute the covariance of a RowMatrix with <= 1 row.")
     val mean = Vectors.fromML(summary.mean)
 
-    if (rows.first().isInstanceOf[DenseVector]) {
+    val sparsityThreshold = 0.5
+    val sparsity = calcSparsity()
+
+    if (sparsity<sparsityThreshold) {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/RowMatrix.scala:444]: Just inline the two vars and put a space around <

**Review by srowen — COMMENTED** [review]
... but I think this is expensive to check? how much difference does it make?

**Comment by srowen** [comment]
Ping @Antozamm

**Pull request closed without merging**